### PR TITLE
 Add module for outputting Kraken-style taxonomy report and --taxLineage

### PR DESCRIPTION
**Pull request by fbreitwieser**
- adds module taxonomyreport for Kraken-style taxonomy report
- add parameter --taxLineage to par.lca for lca, taxonomy and addtaxonomy. Currently this means there is a variable number of fields in the output, which isn't optimal
- refactored NcbiTaxonomy:
  - does not load delnodes.dmp anymore. When a taxon that was seen in a report is not in the taxDB, it could be either through 
  - get rid of two maps; instead D is vector mapping TaxID to internal ID
  - use TaxID type to better distinguish interal IDs and taxIDs

### CODE REVIEW

**Comment by fbreitwieser** [comment]
Changes in mmseqs-benchmark regression to test taxonomyreport:
```sh
# Create taxreport
"${MMSEQS}" taxonomyreport -v 3 "$TARGETDB" "$RESULTS/results_aln" "$RESULTS/results_aln_taxreport"

# Check numbers in taxreport
R_BACTERIA=$(grep 'superkingdom.*Bacteria' "$RESULTS/results_aln_taxreport" | cut -f 2)
R_VIRUS=$(grep 'superkingdom.*Virus' "$RESULTS/results_aln_taxreport" | cut -f 2)
R_EUKARYOTA=$(grep 'superkingdom.*Eukaryota' "$RESULTS/results_aln_taxreport" | cut -f 2)

TARGET="from filtertaxdb: 2524 259 2713; from taxonomyreport: 2524 259 2713"
ACTUAL="from filtertaxdb: $BACTERIA $VIRUS $EUKARYOTA; from taxonomyreport: $R_BACTERIA $R_VIRUS $R_EUKARYOTA"
awk -v actual="$ACTUAL" -v target="$TARGET" 'BEGIN { print (actual == target) ? "GOOD" : "BAD"; \
    print "Expected: ", target; \
    print "Actual:   ", actual; }' > "${RESULTS}/report"
```

File: src/commons/Parameters.cpp [hunk 1]
```diff
@@ -224,16 +224,16 @@ Parameters::Parameters():
...
         PARAM_INVERT_SELECTION(PARAM_INVERT_SELECTION_ID, "--invert", "Invert selection", "Invert selection", typeid(bool), (void*)&invertSelection, ""),
         // view
         PARAM_ID_LIST(PARAM_ID_LIST_ID, "--id-list", "Selected entries with key", "entries to be printed seperated by ','", typeid(std::string), (void*) &idList, ""),
         PARAM_IDX_ENTRY_TYPE(PARAM_IDX_ENTRY_TYPE_ID, "--idx-entry-type", "Index entry type", "sequence; 0, src sequence 1: header: 2, src header :3 (default 0)", typeid(int), (void*) &idxEntryType, "^[0-3]{1}$"),
-
-        // lca
+        // lca and addtaxonomy
         PARAM_LCA_RANKS(PARAM_LCA_RANKS_ID, "--lca-ranks", "LCA ranks", "Ranks to return in LCA computation", typeid(std::string), (void*) &lcaRanks, ""),
         PARAM_BLACKLIST(PARAM_BLACKLIST_ID, "--blacklist", "Blacklisted taxa", "Comma separated list of ignored taxa in LCA computation", typeid(std::string), (void*)&blacklist, "([0-9]+,)?[0-9]+"),
+        PARAM_TAXON_ADD_LINEAGE(PARAM_TAXON_ADD_LINEAGE_ID, "--taxLineage", "Add column with taxon lineage", "Show taxon lineage", typeid(bool), (void*)&showTaxLineage, ""),
```
Comment (milot-mirdita) [inline src/commons/Parameters.cpp]:
```suggestion
        PARAM_TAXON_ADD_LINEAGE(PARAM_TAXON_ADD_LINEAGE_ID, "--tax-lineage", "Show taxon lineage", "Add column with taxon lineage", typeid(bool), (void*)&showTaxLineage, ""),
```

File: src/taxonomy/NcbiTaxonomy.cpp [hunk 2]
```diff
@@ -29,28 +31,36 @@ void deleteMatrix(int** M, size_t maxNodes) {
...
     E.reserve(maxNodes * 2);
     L.reserve(maxNodes * 2);
+
     H = new int[maxNodes];
-    std::fill(H, H + (maxNodes), 0);
-    elh(1, 0);
+    std::fill(H, H + maxNodes, 0);
+
+    std::vector< std::vector<TaxID> > children(taxonNodes.size());
+    for (const TaxonNode& tn : taxonNodes) {
```
Comment (milot-mirdita) [inline src/taxonomy/NcbiTaxonomy.cpp]: Avoid C++11 constructs please

File: src/taxonomy/NcbiTaxonomy.cpp [hunk 3]
```diff
@@ -104,89 +114,45 @@ std::vector<std::string> splitByDelimiter(const std::string &s, const std::strin
...
-        } else {
-            TaxonNode node(parentId, 0, "");
-            node.children.emplace_back(currentId);
-            taxonTree.emplace(parentId, node);
+        TaxID taxId = (TaxID) strtol(result[0].c_str(), NULL, 10);
+        TaxID parentTaxId = (TaxID) strtol(result[1].c_str(), NULL, 10);
+        if (taxId > maxTaxID) {
+            maxTaxID = taxId;
         }
+        taxonNodes.emplace_back(TaxonNode(currentId, taxId, parentTaxId, result[2]));
```
Comment (milot-mirdita) [inline src/taxonomy/NcbiTaxonomy.cpp]:
```suggestion
        taxonNodes.emplace_back(currentId, taxId, parentTaxId, result[2]);
```
Let it automatically forward the parameters with emplace_back

File: src/taxonomy/NcbiTaxonomy.cpp [hunk 4]
```diff
@@ -104,89 +114,45 @@ std::vector<std::string> splitByDelimiter(const std::string &s, const std::strin
...
-void NcbiTaxonomy::restoreRelations() {
-    std::map<int, TaxonNode> helperTree;
-
-    for (std::map<int, TaxonNode>::iterator it = taxonTree.begin(); it != taxonTree.end(); ++it) {
-        TaxonNode node = it->second;
-        node.parentTaxon = D[node.parentTaxon];
-        for (size_t i = 0; i < node.children.size(); ++i) {
-            node.children[i] = D[node.children[i]];
+    // Loop over taxonNodes and check no parents are NULL
+    for (const TaxonNode& tn : taxonNodes) {
```
Comment (milot-mirdita) [inline src/taxonomy/NcbiTaxonomy.cpp]: Same as above, I would avoid the : for loop

File: src/taxonomy/NcbiTaxonomy.h [hunk 5]
```diff
@@ -6,55 +6,63 @@
 #define MMSEQS_NCBITAXONOMY_H
 
 #include <map>
+#include <unordered_map>
 #include <vector>
 #include <string>
 
+using TaxID = int;
```
Comment (milot-mirdita) [inline src/taxonomy/NcbiTaxonomy.h]: Can we do a typedef instead of using?

**Comment by fbreitwieser** [comment]
> Hi Florian,
> 
> Thanks a lot for your pull request. I added some comments. One thing that I am unsure about is the removal of delnodes however. I added that explicitly since an old Uniclust release can have taxids that are already removed in the latest NCBI taxonomy. Then we either get an error or we ignore the protein in Uniprot/Uniclust.

Hi Milot, thansk for looking through it! Re deleted nodes: The reason I think they are not necessary is that for mmseqs it does not matter whether a taxon that is encountered in an report, but is not in the taxonomy database, is from a deleted taxon or a taxon that has been added after the creation of the taxonomy database. In both cases the taxon can be ignored (or an error can be thrown).

File: src/taxonomy/NcbiTaxonomy.h (hunk 5 above)
Comment (fbreitwieser) [reply src/taxonomy/NcbiTaxonomy.h]: Sure. Have a couple of C++11-isms, the others being the foreach loop as you noted. Martin said that he'll discuss them with you whether you want to have them, but as long as it's not auto he is not opposed.

Comment (martin-steinegger) [reply src/taxonomy/NcbiTaxonomy.h]: I noticed that we use 0x for the GNU compiler. I am not sure why C++-11 is active for llvm. We should avoid the foreach loops. Sorry for the confusion.

**Comment by martin-steinegger** [comment]
I think we can leave out the deleted nodes. It makes the code also more clean.

File: src/taxonomy/NcbiTaxonomy.h (hunk 5 above)
Comment (fbreitwieser) [reply src/taxonomy/NcbiTaxonomy.h]: Oh ok, not too much to change. Will fix it.

**Comment by fbreitwieser** [comment]
Btw, what's your policy on double right-angle brackets? E.g. `std::vector<std::vector<int>>`. Wih space or not? Was a problem pre C++11, but maybe works in 0x?
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2005/n1757.html

**Comment by martin-steinegger** [comment]
The regression test finished without errors, however the binary could not be uploaded. Maybe its a problem with the PR. I merged the commit into the master.  
So MMseqs2 is one step closer to be a great taxonomy classification tool. Thanks a lot!

**Pull request merged**